Add multi-tenancy guide (#160)
stolostron/multicluster-mesh-addon: docs/multi-tenancy.md

## Shared Clusters, Separate Control Planes

Multiple meshes on the same ClusterSet must use different `controlPlane.namespace` values. Each mesh operates independently (its certificates, discovery tokens, and trust domain are scoped to its own control plane namespace).

### Prerequisites

A [ManagedClusterSet] with clusters labeled to join it:

```bash
clusteradm create clusterset shared-cluster-set
clusteradm clusterset set shared-cluster-set --clusters cluster1,cluster2
```

### Hub namespaces

Each team gets a dedicated hub namespace.

> **Note:** The add-on controller lists clusters by their ClusterSet label directly. A `ManagedClusterSetBinding` is not required for the add-on to function, though you may want one if using Placements alongside the mesh.

```bash
kubectl create namespace mesh-team-a
kubectl create namespace mesh-team-b
```

### cert-manager trust chain

Each namespace needs its own trust chain (self-signed issuer -> root CA certificate -> CA-backed issuer). The add-on uses the CA-backed issuer to mint intermediate CAs per cluster.

> Apply the trust chain from [`samples/cert-manager-issuer.yaml`](../samples/cert-manager-issuer.yaml) in each team namespace, adjusting the `namespace` and `commonName` fields:

Reviewer: #188 merged, the sample no longer has a hardcoded namespace. The "adjusting the `namespace`" part is now wrong. Merge main and simplify:

```suggestion
Apply the trust chain from [`samples/cert-manager-issuer.yaml`](../samples/cert-manager-issuer.yaml) in each team namespace. Optionally adjust the `commonName` to distinguish each team's root CA (e.g., `Mesh Root CA - Team A`):
```

> ```bash

Reviewer: The `kubectl apply -f ... -n mesh-team-a` command will not place resources into `mesh-team-a` because the sample YAML has explicit `namespace: mesh-system` on each resource. The `-n` flag only sets the default for resources without a namespace.

Two options:

1. Remove the hardcoded namespace from the sample so `-n` works:

```suggestion
Apply the trust chain from [`samples/cert-manager-issuer.yaml`](../samples/cert-manager-issuer.yaml) in each team namespace:

\`\`\`bash
kubectl apply -f samples/cert-manager-issuer.yaml -n mesh-team-a
kubectl apply -f samples/cert-manager-issuer.yaml -n mesh-team-b
\`\`\`
```

This requires removing `namespace: mesh-system` from the sample (separate change).

2. Or make the guide explicit that the user must edit the file first:

```suggestion
Copy `samples/cert-manager-issuer.yaml` and adjust `namespace` to the team namespace and `commonName` to a team-specific value, then apply.
```

Author: Good catch, @mkolesnik. I think option 1 is preferable because hardcoding the namespace in the manifests prevents the -n flag from working, which defeats the purpose of having reusable samples.
Removing the namespace from the manifests makes them reusable across different namespaces and deployment scenarios, such as single-tenant and multi-tenant setups.

I'll open a separate PR to remove the namespace from the manifests and update all the calling contexts.

Author: Done - https://github.com/stolostron/multicluster-mesh-addon/pull/188

## Hub Namespace Isolation

Since `MultiClusterMesh` is namespace-scoped, standard Kubernetes RBAC restricts which teams can create or modify mesh resources. The controller reconciles meshes across all namespaces and enforces conflict rules globally.

Define a `ClusterRole` once and bind it per namespace via `RoleBinding`:

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: mesh-admin
rules:
- apiGroups: ["mesh.open-cluster-management.io"]
  resources: ["multiclustermeshes"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: team-a-mesh-admin
  namespace: mesh-team-a
subjects:
- kind: Group
  name: team-a
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: mesh-admin
  apiGroup: rbac.authorization.k8s.io
```

Repeat the `RoleBinding` for `mesh-team-b` with the appropriate subject.

> In practice, scope full CRUD access to platform engineers and grant the rest of the team read-only (`get`, `list`, `watch`) access.

## Data-Plane Isolation

> The add-on handles control-plane plumbing. Data-plane isolation between co-located meshes requires user-side configuration on each spoke cluster.

Reviewer 2: nit: In my view the controller doesn't do the control plane plumbing for now:
```suggestion
Data-plane isolation between co-located meshes requires user-side configuration on each spoke cluster.
```

# Multi-Tenancy

## Table of Contents

- [Overview](#overview)
- [Shared Clusters, Separate Control Planes](#shared-clusters-separate-control-planes)
- [Hub Namespace Isolation](#hub-namespace-isolation)
- [Data-Plane Isolation](#data-plane-isolation)
- [Conflict Detection and Resolution](#conflict-detection-and-resolution)

## Overview

Multiple `MultiClusterMesh` resources can target the same [ManagedClusterSet], running independent meshes on the same pool of clusters. Isolation between meshes is achieved through:

> 1. **Separate control plane namespaces** - Each mesh installs its Istio control plane in a distinct namespace on spoke clusters

Reviewer: This says each mesh "installs its Istio control plane in a distinct namespace." The add-on installs the operator and mesh plumbing (certs, discovery tokens), but users create the `Istio` CRs themselves (design.md L109, L212). Managing Istio CRs centrally is phase 2 (L244).

Rephrase to something like: "Each mesh uses a separate control plane namespace on spoke clusters" and note that `Istio` CR creation is user-managed.

2. **Namespace-scoped CRD** - `MultiClusterMesh` is namespace-scoped on the hub, enabling RBAC-based tenant boundaries
3. **Per-mesh plumbing** - Each mesh gets its own trust domain, intermediate CA certificates, and discovery tokens

The Sail/OSSM operator is a cluster-scoped singleton shared across meshes. The add-on installs it when the first mesh needs a cluster and removes it only when no mesh targets that cluster anymore.

## Shared Clusters, Separate Control Planes

Multiple meshes on the same ClusterSet must use different `controlPlane.namespace` values. Each mesh operates independently (its certificates, discovery tokens, and trust domain are scoped to its own control plane namespace).

### Prerequisites

A [ManagedClusterSet] with clusters labeled to join it:

```yaml
apiVersion: cluster.open-cluster-management.io/v1beta2
kind: ManagedClusterSet
metadata:
  name: shared-cluster-set
spec:
  clusterSelector:
    selectorType: ExclusiveClusterSetLabel
```

```bash
kubectl label managedcluster cluster1 cluster.open-cluster-management.io/clusterset=shared-cluster-set
kubectl label managedcluster cluster2 cluster.open-cluster-management.io/clusterset=shared-cluster-set

> ```

Reviewer: `clusteradm` has commands for both of these:

```suggestion
clusteradm create clusterset shared-cluster-set
clusteradm clusterset set shared-cluster-set --clusters cluster1,cluster2
```

> ### Hub namespaces

Reviewer: The guide creates hub namespaces for each team but does not mention `ManagedClusterSetBinding`. In OCM, a binding is typically needed to use a ClusterSet from a namespace.

The controller lists clusters by their ClusterSet label directly and does not require a binding. Worth adding a brief note here, e.g.:

> Note: The add-on controller lists clusters by their ClusterSet label directly. A `ManagedClusterSetBinding` is not required for the add-on to function, though you may want one if using Placements alongside the mesh.

Each team gets a dedicated hub namespace:

```bash
kubectl create namespace mesh-team-a
kubectl create namespace mesh-team-b
```

### cert-manager trust chain

> Each namespace needs its own trust chain (self-signed issuer → root CA certificate → CA-backed issuer). The add-on uses the CA-backed issuer to mint intermediate CAs per cluster.

Reviewer: nit: the `→` renders fine in browsers but can be mangled in some terminals. Consider using `->` instead.

Team A (`mesh-team-a`):

> ```yaml

Reviewer: This trust chain duplicates `samples/cert-manager-issuer.yaml` with different names.
You could instead reference the existing sample and note what to change for multi-tenancy:

Apply the trust chain from `samples/cert-manager-issuer.yaml` in each team namespace, adjusting the namespace and commonName.

---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: selfsigned-issuer
  namespace: mesh-team-a
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: mesh-root-ca
  namespace: mesh-team-a
spec:
  isCA: true
  commonName: Team A Mesh Root CA
  secretName: mesh-root-ca-secret
  duration: 87600h
  renewBefore: 720h
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    name: selfsigned-issuer
    kind: Issuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: mesh-root-ca
  namespace: mesh-team-a
spec:
  ca:
    secretName: mesh-root-ca-secret
```

Repeat for `mesh-team-b`.

### MultiClusterMesh resources

> Team A — control plane in `istio-system-team-a`:

Reviewer: nit: em-dashes on lines 102 and 121. Replace with a colon or comma:

```suggestion
Team A, control plane in `istio-system-team-a`:
```

Same on line 121 for Team B.

```yaml
apiVersion: mesh.open-cluster-management.io/v1alpha1
kind: MultiClusterMesh
metadata:
  name: team-a-mesh
  namespace: mesh-team-a
spec:
  clusterSet: shared-cluster-set
  controlPlane:
    namespace: istio-system-team-a
  security:
    trust:
      certManager:
        issuerRef:
          name: mesh-root-ca
```

Team B — control plane in `istio-system-team-b`:

```yaml
apiVersion: mesh.open-cluster-management.io/v1alpha1
kind: MultiClusterMesh
metadata:
  name: team-b-mesh
  namespace: mesh-team-b
spec:
  clusterSet: shared-cluster-set
  controlPlane:
    namespace: istio-system-team-b
  security:
    trust:
      certManager:
        issuerRef:
          name: mesh-root-ca
```

> Both meshes share the operator installation on each cluster. Trust domains (`team-a-mesh`, `team-b-mesh`) and discovery tokens are independent.

Reviewer: nit: both meshes reference `issuerRef.name: mesh-root-ca` which looks like a shared issuer. Worth adding a note that the issuerRef resolves in the mesh's namespace, so each team's `mesh-root-ca` is a different issuer backed by a different root CA.

## Hub Namespace Isolation

Since `MultiClusterMesh` is namespace-scoped, standard Kubernetes RBAC restricts which teams can create or modify mesh resources. The controller reconciles meshes across all namespaces and enforces conflict rules globally.

Example RBAC for Team A:

```yaml

> ---

Reviewer: Two things:

1. The Role is duplicated per namespace. You could define it once as a `ClusterRole` and bind it per namespace via `RoleBinding`. That's the standard pattern for reusable permission sets.

2. This grants every member of `team-a` full CRUD on meshes. Worth a note that in practice you'd scope this to platform engineers and give the rest of the team read-only access.

apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: mesh-admin
  namespace: mesh-team-a
rules:
- apiGroups: ["mesh.open-cluster-management.io"]
  resources: ["multiclustermeshes"]
  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: team-a-mesh-admin
  namespace: mesh-team-a
subjects:
- kind: Group
  name: team-a
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: mesh-admin
  apiGroup: rbac.authorization.k8s.io
```

Repeat for `mesh-team-b` with the appropriate subject.

## Data-Plane Isolation

The add-on handles control-plane plumbing. Data-plane isolation between co-located meshes requires user-side configuration on each spoke cluster.

### `discoverySelectors`

Each Istio control plane discovers services across all namespaces by default. Configure `discoverySelectors` to restrict visibility to the mesh's own application namespaces:

> ```yaml

Reviewer: This section transitions from hub-side resources to spoke-side user configuration without making it clear. Everything above (MultiClusterMesh, Certificates, RBAC) is on the hub.
This `Istio` CR and the namespace labeling below are on each spoke cluster.

Suggestion: add a note at the start of this section, e.g.:

> The following configuration is applied directly on each spoke cluster (not on the hub).

The `kubectl label namespace` commands below (L202-203) need the same clarification.

apiVersion: sailoperator.io/v1
kind: Istio
metadata:
  name: team-a-istio
  namespace: istio-system-team-a
spec:
  values:
    meshConfig:
      discoverySelectors:
      - matchLabels:
          mesh: team-a

> ```

Reviewer: The cert-manager and RBAC sections have "Repeat for mesh-team-b" notes, but this section does not.
Worth adding after L204:

Repeat for Team B with `mesh: team-b`.

Two meshes on the same ClusterSet cannot use the same `controlPlane.namespace`:

```yaml
apiVersion: mesh.open-cluster-management.io/v1alpha1
kind: MultiClusterMesh
metadata:
  name: conflicting-mesh
  namespace: mesh-team-b
spec:
  clusterSet: shared-cluster-set
  controlPlane:
    namespace: istio-system-team-a  # already used by team-a-mesh
  security:
    trust:
      certManager:
        issuerRef:
          name: mesh-root-ca
```

Resulting status on the newer mesh:

```yaml
status:
  conditions:
  - type: Ready
    status: "False"
    reason: NamespaceConflict
    message: 'controlPlane.namespace "istio-system-team-a" conflicts with older mesh
      mesh-team-a/team-a-mesh targeting the same ClusterSet shared-cluster-set'
```

### Operator configuration conflict

Meshes on the same ClusterSet must have compatible operator configurations (channel, source, namespace, etc.) since they share a single operator installation:

```yaml
apiVersion: mesh.open-cluster-management.io/v1alpha1
kind: MultiClusterMesh
metadata:
  name: conflicting-operator-mesh
  namespace: mesh-team-b
spec:
  clusterSet: shared-cluster-set
  controlPlane:
    namespace: istio-system-team-b
  operator:

>     channel: "3.0"  # conflicts with team-a-mesh using "stable"

Reviewer: nit: team-a-mesh does not set `operator.channel` explicitly, so the reader needs to know the kubebuilder default is `"stable"`. Consider:

```suggestion
    channel: "3.0"  # conflicts with team-a-mesh (defaults to "stable")
```
